Api4 CustomGroup.get - use in-memory cache to answer simple calls

ufundo opened the pull request:
Overview
----------------------------------------
@colemanw FYI just wanted to try this.

Before
----------------------------------------
- CustomGroup.get calls always hit the database

After
----------------------------------------
- CustomGroup.get calls that are "simple enough" use the in-memory cache from `CRM_Core_BAO_CustomGroup::getAll`
- otherwise hit the database as normal

Technical Details
----------------------------------------
I fear there might be some subtle differences.

Indeed, I have already found one around case sensitivity.
- DAO select queries seem to be at least partially case insensitive: `->where('name', '=', 'abc')` returns a result where the group name actually equals 'Abc'
- using the cache it becomes case-sensitive

This feels like an oddity of Api4 in general though - where is functioning slightly differently for DAOGetActions vs BasicGetActions?

ufundo committed:
Api4 CustomGroup.get - use in-memory cache to answer simple calls

ufundo commented:
Just noting this reversion if this got to mergeable: https://github.com/ufundo/civicrm-core/commit/056572c86b35639f5bb5563102ff5ebacb9bf631

I also just grepped `civicrm-core` and there are a handful of other examples of simple CustomGroup / CustomField gets that look like they would benefit. I daresay more out there in the world/universe

ufundo commented:
I also wondered if something this could become a general Trait for other Api4 entities which have a small number of records.

colemanw commented:
@ufundo perhaps this would solve the case-sensitivity issue:
```diff
--- a/Civi/Api4/Generic/Traits/ArrayQueryActionTrait.php
+++ b/Civi/Api4/Generic/Traits/ArrayQueryActionTrait.php
@@ -127,6 +127,13 @@ trait ArrayQueryActionTrait {
       case '=':
       case '!=':
       case '<>':
+        // For parity with SQL operators, do case-insensitive matching
+        if (is_string($value)) {
+          $value = strtolower($value);
+        }
+        if (is_string($expected)) {
+          $expected = strtolower($expected);
+        }
         $equal = $value == $expected;
         // PHP is too imprecise about comparing the number 0
         if ($expected === 0 || $expected === '0') {
```

colemanw commented:
> I also wondered if something this could become a general Trait for other Api4 entities which have a small number of records.

Custom fields would be the obvious 2nd candidate.

Custom groups & fields are cached due to their ubiquity throughout the system. Not sure if there are other entities like that... contact types perhaps?

colemanw commented on Civi/Api4/Action/CustomGroup/Get.php:
```diff
@@ -0,0 +1,84 @@
+   * This acts like a BasicBatchAction - ie provide all the records
+   * upfront, and then filter using queryArray
+   *
+   * NOTE: we skip formatValues because any pseudoconstant fields
+   * will trigger 'needDb'
```
The `BasicGetAction` is designed to handle pseusoconstants, so I think those could be included in the "use cache" camp.

ufundo replied:
I avoided pseudoconstant suffixes because I think they need the `formatValues` function from BasicGetAction, and that looked non-trivial to port across.

colemanw commented on Civi/Api4/Action/CustomGroup/Get.php:
```diff
@@ -0,0 +1,84 @@
+      return TRUE;
+    }
+    $standardFields = \Civi::entity('CustomGroup')->getFields();
+    foreach ($this->select as $field) {
```
To include pseudoconstants:
```suggestion
    foreach ($this->select as $field) {
      // Strip pseudoconstant suffix
      [$field] = explode(':', $field);
```

colemanw commented:
@ufundo here's a PR for the case-sensitivity thing: https://github.com/civicrm/civicrm-core/pull/31514

ufundo commented on Civi/Api4/Action/CustomGroup/Get.php:
```diff
@@ -0,0 +1,84 @@
+    parent::getObjects($result);
+  }
+
+  protected function needDb() {
```
> Clauses that use field-to-field comparisons

Do you mean they don't technically "need" the DB? That makes sense but at this stage I would be happy to leave those out of scope because it like a chunk more work for minimal gain?

ufundo committed:
CustomGroup.get - check where and orderBy fields are answerable from cache

ufundo committed:
CustomGroup.get - add pseudoconstant field handling

ufundo committed:
CustomGroup.get - improve comments, function names

ufundo commented:
> Custom groups & fields are cached due to their ubiquity throughout the system. Not sure if there are other entities like that... contact types perhaps?

I was wondering about OptionGroups and OptionValues - though it might be harder to integrate seamlessly with existing pseudoconstant cache?

I had a go for CustomField https://github.com/ufundo/civicrm-core/pull/7/files

colemanw commented on Civi/Api4/Action/CustomGroup/Get.php:
```diff
@@ -0,0 +1,103 @@
+      [$field] = explode(':', $clause[0] ?? '');
+      if (!$field || !isset($standardFields[$field])) {
+        return TRUE;
+      }
```
@ufundo to leave the field-to-field comparisons out of scope we just need to exclude them here, like:
```suggestion
      }
      // ArrayQueryTrait doesn't yet support field-to-field comparisons
      if (!empty($clause[3]) {
        return TRUE;
      }
```

eileenmcnaughton commented:
test this please

ufundo committed:
Api4 CustomGroup cache - exclude field-to-field comparisons

Co-authored-by: colemanw <coleman@civicrm.org>

ufundo committed:
add test for CustomGroup.get caching

ufundo commented:
Fixed the github commit snafu.

I've also added a test. 

It includes a case-sensitivity case that I expect to fail until https://github.com/civicrm/civicrm-core/pull/31508 is merged - **this should not be merged before that.**

colemanw commented:
@ufundo the only other monkey-wrench is that the `needsDb` function would need to recursively walk through nested and/or/not groups in the WHERE clause. See `\Civi\Api4\Query\Api4Query::treeWalkClauses`

ufundo commented:
Oh, another good point....

At the moment won't they be caught by the check for known field names? Unless you have a field on your entity named `AND`. Though that probably isn't recommended for a number of reasons. :roll_eyes:

ufundo commented on Civi/Api4/Action/CustomGroup/Get.php:
```diff
@@ -0,0 +1,107 @@
+    }
+    foreach ($this->where as $clause) {
+      [$field] = explode(':', $clause[0] ?? '');
+      if (!$field || !isset($standardFields[$field])) {
+        return TRUE;
```
@colemanw 'AND' 'OR' etc wont be set in $standardFields ?

colemanw commented:
> At the moment won't they be caught by the check for known field names?

Ah, I see what you're saying. Yes that's true. Using AND/OR/NOT would trigger the fallback to SQL (even though ArrayQuery actually can handle those).

> Unless you have a field on your entity named AND. Though that probably isn't recommended for a number of reasons. 🙄

Um, yea. Api4 does an all-round good job of avoiding name collisions, but no, you shouldn't name your table columns AND or OR or NOT. That would crash everything, so don't do that.

ufundo commented:
Noted :)

colemanw commented:
Also, just to be safe, I wouldn't recommend naming a column "DROP TABLE civicrm_contacts;"

ufundo commented:
Retest this please

ufundo commented:
Tests passing thanks to @colemanw 's case sensitivity PR :raised_hands:

colemanw commented:
Great work @ufundo. Merged. Wanna put up the one for CustomField?

ufundo commented:
:grin:  https://github.com/civicrm/civicrm-core/pull/31569
